src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -751,3 +859,275 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+	{
+		prev_shmem_startup_hook();
+	}
+}
+
+
+/*
+ * RouterTransactionCallback handles committing or aborting remote transactions
+ * after the local one has committed or aborted. It only sends COMMIT or ABORT
+ * commands to still-healthy remotes; the failed ones are marked as inactive if
+ * after a successful COMMIT (no need to mark on ABORTs).
+ */
+static void
+RouterTransactionCallback(XactEvent event, void *arg)
+{
+	if (xactParticipantHash == NULL)
+	{
+		return;
+	}
+
+	switch (event)
+	{
+#if (PG_VERSION_NUM >= 90500)
+		case XACT_EVENT_PARALLEL_COMMIT:
+#endif
+		case XACT_EVENT_COMMIT:
+		{
+			bool commit = true;
+
+			ExecuteTransactionEnd(commit);
+
+			break;
+		}
+
+#if (PG_VERSION_NUM >= 90500)
+		case XACT_EVENT_PARALLEL_ABORT:
+#endif
+		case XACT_EVENT_ABORT:
+		{
+			bool commit = false;
+
+			ExecuteTransactionEnd(commit);
+
+			break;
+		}
+
+		/* no support for prepare with long-running transactions */
+		case XACT_EVENT_PREPARE:
+		case XACT_EVENT_PRE_PREPARE:
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot prepare a transaction that modified "
+								   "distributed tables")));
+
+			break;
+		}
+
+#if (PG_VERSION_NUM >= 90500)
+		case XACT_EVENT_PARALLEL_PRE_COMMIT:
+#endif
+		case XACT_EVENT_PRE_COMMIT:
+		{
+			if (subXactAbortAttempted)
+			{
+				subXactAbortAttempted = false;
+
+				ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+								errmsg("cannot ROLLBACK TO SAVEPOINT in transactions "
+									   "which modify distributed tables")));
+			}
+
+			/* leave early to avoid resetting transaction state */
+			return;
+		}
+	}
+
+	/* reset transaction state */
+	hash_destroy(xactParticipantHash);
+	xactParticipantHash = NULL;
+	subXactAbortAttempted = false;
+}
+
+
+/*
+ * pgfdw_subxact_callback --- cleanup at subtransaction end.
+ */
+static void
+RouterSubtransactionCallback(SubXactEvent event, SubTransactionId subId,
+							 SubTransactionId parentSubid, void *arg)
+{
+	if ((xactParticipantHash != NULL) && (event == SUBXACT_EVENT_ABORT_SUB))
+	{
+		/* TODO: warn here */
+		subXactAbortAttempted = true;
+	}
+}
+
+
+/* TODO: Write comment */
+static HTAB *
+CreateXactParticipantHash(void)
+{
+	HTAB *xactParticipantHash = NULL;
+	HASHCTL info;
+	int hashFlags = 0;
+
+	MemSet(&info, 0, sizeof(info));
+	info.keysize = sizeof(XactParticipantKey);
+	info.entrysize = sizeof(XactParticipantEntry);
+	info.hcxt = TopTransactionContext;
+	hashFlags = (HASH_ELEM | HASH_CONTEXT);
+
+#if (PG_VERSION_NUM >= 90500)
+	hashFlags |= HASH_BLOBS;
+#else
+	hashFlags |= HASH_FUNCTION;
+	info.hash = tag_hash;
+#endif
+
+	xactParticipantHash = hash_create("citus xact participant hash", 32, &info,
+									  hashFlags);
+
+	return xactParticipantHash;
+}
+
+
+/* Allocates eight bytes, and copies given value's contents those bytes. */
+static uint64 *
+AllocateUint64(uint64 value)
+{
+	uint64 *allocatedValue = (uint64 *) palloc0(sizeof(uint64));
+	Assert(sizeof(uint64) >= 8);
```

Comment:
Copy-pasta. Not going to change, since I don't remember writing it. I'll get rid of my copy-pasted function in my clean-up commit.